### PR TITLE
Sharded gossip tests

### LINKED ISSUE #3436: Tests for sharded gossip
Ideally using the same mocks that are being used by the most recent Kitsune integration tests. We need to find a way to write tests for sharded gossip. There currently aren't any tests that actually check what happens when sharding kicks in. There are tests that verify the behaviour of gossip with fewer nodes, just not the sharding cases. A potential approach could be to turn the sharding threshold down so that it kicks in during a test with 5/6 nodes because we can only run so many instances of Kitsune reliably in a single test on CI.

### PR DESCRIPTION
**Pull request by ThetaSinner**
### Summary

Closes #3436

### TODO:
- [ ] CHANGELOGs updated with appropriate info

### CODE REVIEW

File: crates/kitsune_p2p/dht/src/arq.rs [hunk 1]
```diff
@@ -580,7 +581,7 @@ pub fn power_and_count_from_length(dim: impl SpaceDim, len: u64, max_chunks: u32
     let dim = dim.get();
     assert!(len <= U32_LEN);
     let mut power = 0;
-    let mut count = (len / dim.quantum as u64) as f64;
+    let mut count = (len as f64 / dim.quantum as f64).ceil();
```
Comment (ThetaSinner) [inline crates/kitsune_p2p/dht/src/arq.rs:584]: Division was rounding down here which always resulted in an arq that was up to 1 quantum smaller than the arc in standard coordinates. This had strange behaviour for small standard arcs (less than the quantum size) where the arc would be mapped to empty rather than the quantum size. That seems wrong to me, the arq does not represent the arc if it loses part of the arc space. Especially for a small arc, it seems wrong for a small arc to become empty when it wasn't empty. Empty is really a special value.

File: crates/kitsune_p2p/dht/src/arq/arq_set.rs [hunk 2]
```diff
@@ -305,6 +304,71 @@ mod tests {
         );
     }
 
+    #[test]
+    fn arq_set_is_union() {
```
Comment (ThetaSinner) [inline crates/kitsune_p2p/dht/src/arq/arq_set.rs:308]: This test doesn't pass without the change above

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/initiate.rs [hunk 3]
```diff
@@ -67,6 +68,7 @@ impl ShardedGossipLocal {
                 inner.initiate_tgt = Some(tgt);
                 Ok(())
             })?;
+            // TODO always a new connection even if we have an open one?
```
Comment (ThetaSinner) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/initiate.rs:71]: Missed this previously, if we gossip with somebody more than once then shouldn't we be re-using the connection if we can?

File: crates/kitsune_p2p/kitsune_p2p/tests/common/test_legacy_host.rs [hunk 4]
```diff
@@ -268,11 +268,20 @@ impl TestLegacyHost {
                                         for interval in intervals {
                                             match interval {
                                                 DhtArcRange::Bounded(lower, upper) => {
-                                                    if lower < op.location()
-                                                        && op.location() < upper
-                                                    {
-                                                        in_any = true;
-                                                        break;
+                                                    if lower <= upper {
```
Comment (ThetaSinner) [inline crates/kitsune_p2p/kitsune_p2p/tests/common/test_legacy_host.rs:271]: Incomplete test host implementation. The real host already checks the bounded arc this way

File: crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs [hunk 5]
```diff
@@ -42,120 +45,134 @@ async fn publish_to_basis_from_inside() {
...
-            let mut agent = harness.create_agent().await;
-            let mut found_loc = false;
-            for _ in 0..1000 {
-                let loc = agent.get_loc().as_();
-                // Search from the start of our agent, up to halfway through it. Agents that are in the
-                // upper part of their range are less likely to overlap with the previous agent.
-                if (loc) > base_len * i && (loc as f64) < base_len as f64 * (i as f64 + 0.5) {
-                    found_loc = true;
-                    break;
+    let agents = setup_overlapping_agents(
```
Comment (ThetaSinner) [inline crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs:51]: Existing tests haven't changed, I've just extracted some code to reduce duplication

File: crates/kitsune_p2p/dht/src/arq/arq_set.rs [hunk 6]
```diff
@@ -305,6 +304,95 @@ mod tests {
...
+        let dim = SpaceDimension::standard();
+        let strat = ArqStrat::default();
+
+        let start_a = 10_000;
+        let len_a = 30_000;
+        let arq_a =
+            Arq::from_start_and_half_len_approximate(dim, &strat, start_a.into(), len_a / 2);
+
+        let start_b = 10_000_000;
+        let len_b = 20_000;
```
Comment (maackle) [inline crates/kitsune_p2p/dht/src/arq/arq_set.rs:318]: These are extremely small arcs. The quantum size is 2^12 = 4096. For perspective, the quantum size was selected to support sharded networks with millions of connected nodes ([justification](https://github.com/holochain/holochain/blob/c5480bea43b0810b395a3fdd2e24653795aa76b2/crates/kitsune_p2p/dht/src/spacetime/topology.rs#L162-L176)). So while it's good to test at arbitrary scales, this is a scale we won't be seeing any time soon, or perhaps ever.

File: crates/kitsune_p2p/dht/src/arq/arq_set.rs [hunk 7]
```diff
@@ -305,6 +304,95 @@ mod tests {
...
+            let agent_arc_overlap_start_a = Arq::from_start_and_half_len_approximate(
+                dim,
+                &strat,
+                (start_a - 1_000).into(),
+                1_500,
+            );
+            let interval = DhtArcRange::from(agent_arc_overlap_start_a.to_dht_arc_std());
+
+            assert!(arc_set.overlap(&interval.into()));
+        }
```
Comment (maackle) [inline crates/kitsune_p2p/dht/src/arq/arq_set.rs:345]: `100` and `1500` are both sub-quantum values, so it's misleading to use different values which round to the same thing (4096). I think this could be a more useful test if it used larger scales which comprise many quanta.

File: crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs (hunk 5 above)
Comment (maackle) [reply crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs:51]: Are you saying that we don't really need to review what follows in this file?

File: crates/kitsune_p2p/dht/src/arq.rs (hunk 1 above)
Comment (ThetaSinner) [reply crates/kitsune_p2p/dht/src/arq.rs:584]: Okay, that sounds like something that'd be better off on its own PR

Comment (ThetaSinner) [reply crates/kitsune_p2p/dht/src/arq.rs:584]: Moved to an issue https://github.com/holochain/holochain/issues/3976

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/initiate.rs (hunk 3 above)
Comment (ThetaSinner) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/initiate.rs:71]: Ahh right! I'll get rid of this TODO then. Thank you

File: crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs (hunk 5 above)
Comment (ThetaSinner) [reply crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs:52]: The existing tests haven't changed, but the PR is adding new tests below. Those are important and I do need a review for them please. Was just trying to help out with this diff at the top of the file - wasn't clear enough!

File: crates/kitsune_p2p/dht/src/arq/arq_set.rs (hunk 7 above)
Comment (ThetaSinner) [reply crates/kitsune_p2p/dht/src/arq/arq_set.rs:345]: I understand the point that smaller networks will have far larger arcs than this, yes I agree. I don't see why it follows that unit testing should avoid testing at this scale. My aim here was to pick numbers that I can easily read off and see the logic. If i use calculations or much larger numbers then I'm not able to do that.

Using sub-quantum values helped reveal an issue with small values where they are collapsed to 0 rather than expanded to 1 quanta. Even if the code I found the problem in is test-only, I still think that helps justify this test.

I'm also thinking that testing at an arbitrary scale like this should generalise. If this works, then larger arcs and more arcs should also work.

Comment (maackle) [reply crates/kitsune_p2p/dht/src/arq/arq_set.rs:345]: I agree with your reasoning, but in at least one way this is a special case that doesn't generalize:

The arc resizing algo always produces arcs that are between 8 and 16 segments in size, and the minimum segment size is of course 1 quantum. So, for the entire arc to be only 1 quantum long means we have a single-segment arc, which is not typical. Technically the algo should go all the way down to 1 quantum segment if it needs to, but that is the only time an arc will contain only 1 segment.

It's certainly OK to test this, but I think the usefulness is more in that it *is* an extreme edge case. For testing virtually all other cases, we at least need arcs that are 8-16 segments long.

Comment (ThetaSinner) [reply crates/kitsune_p2p/dht/src/arq/arq_set.rs:345]: Understood, yes that's fair that in the context that these types are used these tests are not realistic. I was trying to understand the pieces in isolation and I'm happy with these being an extreme case in that context.

File: crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs (hunk 5 above)
Comment (maackle) [reply crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs:54]: @ThetaSinner it's hard to tell what's new and what's changed from the diff, could you point out which tests are new?

File: crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs [hunk 8]
```diff
@@ -193,46 +215,131 @@ async fn publish_to_basis_from_inside() {
...
+    assert_eq!(1, agents[should_recv_idx].0.op_store().read().len());
+
+    check_op_receivers(&agents, basis, &[sender_idx, should_recv_idx]);
 }
 
-/// Very similar to the test above except the publisher does not have the basis in its arc.
-/// This is a valid scenario because any hash might be produced by creating data and the publish
-/// should still go to the correct agents. It also stays with the publisher, so we need to account
-/// for that when checking the op stores at the end of the test.
+/// Similar to the test above except that we create the op for an agent outside the arc that is belongs in.
```
Comment (ThetaSinner) [inline crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs:334]:
```suggestion
/// Similar to the test above except that we create the op for an agent outside the arc that it belongs in.
```

File: crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs (hunk 5 above)
Comment (ThetaSinner) [reply crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs:54]: Oh right, yeah that diff is really unclear because the common setup I extracted has got mixed in with the last test.

The new tests are `gossip_to_basis_from_inside` and `gossip_to_basis_from_outside`

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/initiate.rs [hunk 9]
```diff
@@ -49,6 +49,7 @@ impl ShardedGossipLocal {
         {
             let id = rand::thread_rng().gen();
 
+            // TODO Why send both the agents and the intervals? The agents contain their arcs
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/initiate.rs:52]: good question

File: crates/kitsune_p2p/kitsune_p2p/tests/common/test_legacy_host.rs [hunk 10]
```diff
@@ -268,8 +268,15 @@ impl TestLegacyHost {
...
-                                                        && op.location() < upper
+                                                    if lower <= upper {
+                                                        if lower < op.location()
+                                                            && op.location() < upper
+                                                        {
+                                                            in_any = true;
+                                                            break;
+                                                        }
+                                                    } else if lower < op.location()
+                                                        || op.location() < upper
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/tests/common/test_legacy_host.rs:279]: should the bounds be inclusive?

**Review by maackle — APPROVED** [review]

File: crates/kitsune_p2p/kitsune_p2p/tests/common/test_legacy_host.rs (hunk 10 above)
Comment (ThetaSinner) [reply crates/kitsune_p2p/kitsune_p2p/tests/common/test_legacy_host.rs]: Yes in the first case and no in the else-if https://github.com/holochain/holochain/blob/develop/crates/holochain/src/conductor/space.rs#L431

Or to put it another way, yes it's inclusive and when describing a 'wrapping' inclusive the correct way to check that is with non-inclusive checks.
